pkg/core/blockchain.go

```diff
@@ -314,6 +318,82 @@ func (bc *Blockchain) Run() {
 	}
 }
 
+// notificationDispatcher manages subscription to events and broadcasts new events.
+func (bc *Blockchain) notificationDispatcher() {
+	var (
+		// These are just sets of subscribers, though modelled as maps
+		// for ease of management (not a lot of subscriptions is really
+		// expected, but maps are convenient for adding/deleting elements).
+		blockFeed        = make(map[chan<- *block.Block]bool)
+		txFeed           = make(map[chan<- *transaction.Transaction]bool)
+		notificationFeed = make(map[chan<- *state.NotificationEvent]bool)
+		executionFeed    = make(map[chan<- *state.AppExecResult]bool)
+	)
+	for {
+		select {
+		case <-bc.stopCh:
+			return
+		case sub := <-bc.subCh:
+			switch ch := sub.(type) {
+			case chan<- *block.Block:
+				blockFeed[ch] = true
+			case chan<- *transaction.Transaction:
+				txFeed[ch] = true
+			case chan<- *state.NotificationEvent:
+				notificationFeed[ch] = true
+			case chan<- *state.AppExecResult:
+				executionFeed[ch] = true
+			default:
+				panic(fmt.Sprintf("bad subscription: %T", sub))
+			}
+		case unsub := <-bc.unsubCh:
+			switch ch := unsub.(type) {
+			case chan<- *block.Block:
+				delete(blockFeed, ch)
+			case chan<- *transaction.Transaction:
+				delete(txFeed, ch)
+			case chan<- *state.NotificationEvent:
+				delete(notificationFeed, ch)
+			case chan<- *state.AppExecResult:
+				delete(executionFeed, ch)
+			default:
+				panic(fmt.Sprintf("bad unsubscription: %T", unsub))
+			}
+		case event := <-bc.events:
+			// We don't want to waste time looping through transactions when there are no
+			// subscribers.
+			if len(txFeed) != 0 || len(notificationFeed) != 0 || len(executionFeed) != 0 {
+				var aerIdx int
+				for _, tx := range event.block.Transactions {
+					if tx.Type == transaction.InvocationType {
+						aer := event.appExecResults[aerIdx]
+						if !aer.TxHash.Equals(tx.Hash()) {
+							panic("inconsistent application execution results")
+						}
+						aerIdx++
+						for ch := range executionFeed {
+							ch <- aer
+						}
+						if aer.VMState == "HALT" {
+							for i := range aer.Events {
+								for ch := range notificationFeed {
+									ch <- &aer.Events[i]
+								}
+							}
+						}
+					}
+					for ch := range txFeed {
+						ch <- tx
+					}
+				}
+			}
+			for ch := range blockFeed {
+				ch <- event.block
```

Comment:
Maybe it is better to use non-blocking calls in each of this sends? It seems wrong to hang here, because `storeBlock` will hang too.

Reply:
Well, there is question of whether we can or cannot lose these events. Losing events is obviously bad, but there are cases where we can't ensure proper event processing on the receiving side like for networked clients, so we have to allow event skipping and deal with that. Blockchain-level subscriptions on the other hand are internal to the process using Blockchain, so not handling events here is a programmer error, IMO. And yes, this error has some consequences that are described in the documentation.